File: src/decoders/meal.ts
import { type Meal, DishKind } from "~/models";
import { decodeFood } from "./food";

export const decodeMeal = (meal: any): Meal => {
  const dishes = meal.ListePlats.V.reduce((acc: any, dish: any) => {
    let key: string;

    switch (dish.G) {
      case DishKind.Entry:
        key = "entry";
        break;
      case DishKind.Main:
        key = "main";
        break;
      case DishKind.Side:
        key = "side";
        break;
      case DishKind.Drink:
        key = "drink";
        break;
      case DishKind.Fromage:
        key = "fromage";
        break;
      case DishKind.Dessert:
        key = "dessert";
        break;
      default:
        throw new Error("unreachable");
    }

    acc[key] = dish.ListeAliments.V.map(decodeFood);
    return acc;
  }, {});

  return {
    name: meal.L,
    ...dishes
  };
};

File: library/src/commonMain/kotlin/decoders/meal.kt
package decoders

import kotlinx.serialization.json.*
import models.DishKind
import models.Food
import models.Meal

fun decodeMeal (meal: JsonObject): Meal {
    var entry: List<Food>? = null
    var main: List<Food>? = null
    var side: List<Food>? = null
    var drink: List<Food>? = null
    var fromage: List<Food>? = null
    var dessert: List<Food>? = null

    meal["ListePlats"]!!.jsonObject["V"]!!.jsonArray.map {dish ->
        val data = dish.jsonObject["ListeAliments"]!!.jsonObject["V"]!!.jsonArray.map { decodeFood(it.jsonObject) }

        when (DishKind.fromInt(dish.jsonObject["G"]!!.jsonPrimitive.int)) {
            DishKind.Entry -> entry = data
            DishKind.Main -> main = data
            DishKind.Side -> side = data
            DishKind.Drink -> drink = data
            DishKind.Fromage -> fromage = data
            DishKind.Dessert -> dessert = data
        }
    }

    return Meal(
        name = meal["L"]?.jsonPrimitive?.content,
        entry = entry,
        main = main,
        side = side,
        drink = drink,
        fromage = fromage,
        dessert = dessert
    )
}